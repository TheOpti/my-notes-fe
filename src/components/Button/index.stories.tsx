import * as React from 'react';
import { storiesOf } from '@storybook/react';

import Button from './';

const btnWrapperStyle = {
  margin: '20px',
};

storiesOf('Button', module)
  .add('all buttons', (): any => (
    <>
      <div style={btnWrapperStyle}>
        <Button label="Raised button" color="raised" />
      </div>
      <div style={btnWrapperStyle}>
        <Button label="Outlined button" color="outlined" />
      </div>
      <div style={btnWrapperStyle}>
        <Button label="Raised with loading button" loading color="raised" />
      </div>
      <div style={btnWrapperStyle}>
        <Button label="Outlined with loading button" loading color="outlined" />
      </div>
      <div style={btnWrapperStyle}>
        <Button label="Disabled raised button" disabled color="raised" />
      </div>
      <div style={btnWrapperStyle}>
        <Button label="Disabled outlined button" disabled color="outlined" />
      </div>
    </>
  ))
  .add('with text', () => (
    <Button label="This is sample button" />
  ));