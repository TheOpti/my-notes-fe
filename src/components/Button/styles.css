.root {
  border: none;
  border-radius: 4px;
  padding: 0 16px;
  min-width: 64px;
  min-height: 36px;
  outline: none;
  cursor: pointer;
  font-size: 16px;
  transition: box-shadow 0.3s, opacity 0.3s, background-color 0.3s;
}

.content {
  display: flex;
  align-items: center;
  justify-content: center;
}

.raised {
  background-color: var(--main-blue-color);
  color: white;
}

.raised:not(.disabled):hover {
  box-shadow: 0 2px 4px -1px rgba(0, 0, 0, 0.2);
  opacity: 0.9;
}

.outlined {
  background-color: white;
  box-shadow: none;
  color: var(--main-blue-color);
  border: 1px solid currentColor;
}

.outlined:not(.disabled):hover {
  background-color: rgba(33, 150, 243, 0.1);
}

.disabled {
  opacity: 0.8;
  cursor: not-allowed;
}