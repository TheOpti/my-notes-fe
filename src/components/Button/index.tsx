import React from 'react';
import cx from 'classnames';

import LoadingSpinner from '../LoadingSpinner';

import styles from './styles.css';

type ButtonPropsType = {
  label: string;
  onClickHandler?: (event: any) => void;
  color?: 'raised' | 'outlined';
  classname?: string,
  disabled?: boolean,
  loading?: boolean,
}
const Button: React.SFC<ButtonPropsType> = (props: ButtonPropsType) => {
  const { label, onClickHandler, color, classname, disabled, loading } = props;

  const btnClasses = cx(styles.root, styles[color], classname, {
    [styles.disabled]: disabled || loading,
  });

  const labelStyles = {
    ...(loading ? { paddingLeft: '8px' }: null),
  };

  return (
    <button 
      onClick={onClickHandler}
      className={btnClasses}
      type="button"
    >
      <span className={styles.content}>
        { loading && <LoadingSpinner size="small" monocolor />}
        <span style={labelStyles}>
          { label }
        </span>
      </span>
    </button>
  );
}

Button.defaultProps = {
  label: '',
  color: 'raised',
  classname: '',
};

export default Button;
