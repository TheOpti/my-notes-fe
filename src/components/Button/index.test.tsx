import React from "react";
import { mount } from 'enzyme';
import Button from './';

const noop = () => {};

describe('Button', () => {
  it('should render correctly without any issues', () => {
    const wrapper = mount(<Button 
      label="Label"
      onClickHandler={noop}
    />);

    expect(wrapper).toMatchSnapshot();
  });
  
  it('should call onClickHandler function when button is actually clicked', () => {
    const mockFunction = jest.fn();

    const wrapper = mount(<Button 
      label="Label"
      onClickHandler={mockFunction}
      disabled
    />);

    wrapper.find('button').simulate('click');
    expect(mockFunction).toHaveBeenCalled();
  });

  it('should render disabled button when "disabled" property is set to true', () => {
    const wrapper = mount(<Button 
      label="Label"
      onClickHandler={noop}
      disabled
    />);

    const containsDisabled = wrapper.find('button').html().includes('disabled');
    expect(containsDisabled).toBe(true);
  });
});