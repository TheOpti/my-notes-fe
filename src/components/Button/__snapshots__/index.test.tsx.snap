// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render correctly without any issues 1`] = `
<Button
  classname=""
  color="raised"
  label="Label"
  onClickHandler={[Function]}
>
  <button
    className="root raised"
    onClick={[Function]}
    type="button"
  >
    <span
      className="content"
    >
      <span
        style={Object {}}
      >
        Label
      </span>
    </span>
  </button>
</Button>
`;
